### ISSUE TITLE
Sphere coordinates should be fixed

### ISSUE THREAD

**Issue by MarkJeronimus** [openscad/openscad#462]
A common [CG sphere](https://www.google.com/search?tbm=isch&q=wireframe+sphere) has coordinates at integral positions along the longitude and latitude axes, but OpenSCAD spheres have vertices on the latitude axis on half positions instead.

To elaborate, if $fn=12, the sphere has longitude vertices at 0°, 30°, 60°, etc. but on the latitude, the vertices are at 15°, 45°, 75°, etc. creating an unwanted <i>flat</i> top and bottom.

This creates scenarios where geometry doesn't match with other geometry, or geometry ends up too thin. The full discussion with examples is on http://forum.openscad.org/Fix-the-sphere-with-vertex-on-top-td5332.html

Temporary workaround (rotation of a half-circle):

<pre><code>module my_sphere() {
    rotate_extrude()
    intersection() {
        circle();
        translate([0, -1]) square(2);
    }
}</code></pre>

(note that a rotation of a full circle causes a void object when rendering F6).
##

**Comment by thehans** [comment]
I thought I would post my version of this sphere implementation.  

By generating list of points for a semicircle and rotate_extrude'ing that, you save a boolean operation.  And you also don't have to throw away half of the circle points you just computed.

This uses the arc, and fragments functions that I orignally developed for my "[Functional OpenSCAD](https://github.com/thehans/FunctionalOpenSCAD)" library which operates directly on vertex data.  

```
module my_sphere(r=1, d) { 
  let(R = d == undef ? r : d/2)
  rotate_extrude()
    polygon(arc(r=R, angle=180, offsetAngle=-90));
}

// Draw a circular arc with center c, radius r, etc.
// "center" parameter centers the sweep of the arc about the offsetAngle (half to each side of it)
// "internal" parameter enables polyhole radius correction
function arc(r=1, angle=360, offsetAngle=0, c=[0,0], center=false, internal=false) = 
  let (
    fragments = ceil((abs(angle) / 360) * fragments(r,$fn)),
    step = angle / fragments,
    a = offsetAngle-(center ? angle/2 : 0),
    R = internal ? r / cos (180 / fragments) : r,
    last = (abs(angle) == 360 ? 1 : 0)
  )
  [ for (i = [0:fragments-last] ) let(a2=i*step+a) c+R*[cos(a2), sin(a2)] ];

function fragments(r=1) = ($fn > 0) ? 
  ($fn >= 3 ? $fn : 3) : 
  ceil(max(min(360.0 / $fa, r*2*PI / $fs), 5));

```

**Comment by espr14** [comment]
This produces more precise sphere with the same number of triangles. There are even more precise triangulations but this keeps the points on the axes (if $fn % 4 == 0).

![sphere](https://user-images.githubusercontent.com/46957311/82740507-9d923680-9d49-11ea-80c0-63058ad06ec7.png)

```
module sphere(r=1,$fn=$fn){
	n=$fn/2+1;
	efn=2*ceil($fn/2);

	points = concat([[0,0,-r]],
		[for(j=[1:efn/2], i=[0:$fn-1])
			[r*cos((j-efn/4)*360/efn)*cos((i-j/2)*360/$fn),r*cos((j-efn/4)*360/efn)*sin((i-j/2)*360/$fn),r*sin((j-efn/4)*360/efn)]],
		[[0,0,r]]
	);

	faces = concat(
		 ///bottom
		[for(i=[0:$fn-2]) [0,i+1,i+2]], [[0,$fn,1]],
			
		///even on side
		[for(z=[1:$fn:$fn*(efn/2-2)+1])
			each concat(
				[for(i=[0:$fn-2]) [z+i,z+$fn+i,z+$fn+i+1]],
				[[z+$fn-1,z+$fn+$fn-1,z+$fn]]
			)
		],

		///odd on side
		[for(z=[1:$fn:$fn*(efn/2-2)+1])
			each concat(
				[for(i=[0:$fn-2]) [z+i,z+$fn+i+1,z+i+1]],
				[[z+$fn-1,z+$fn,z]]
			)
		]
		);


	polyhedron(points=points, faces=faces);
}
```

**Comment by jordanbrown0** [comment]
Related to #1990, which is arguably the 2D case of the same thing:  when the vertices aren't on the axes, the bounding box is wrong.

**Comment by nophead** [comment]
Yes I have simply redefined sphere() in NopSCADlib to have points on all six half axes. I have never found a problem with doing that.

**Comment by jordanbrown0** [comment]
#1983 seems to have more discussion.
